code_quality.js

```diff
@@ -0,0 +1,337 @@
+         return;
+      }
+      console.log('File to be analyzed before commit:\n' + files.join('\n'));
+      filesByExtension = groupFilesByExtension(files);
+   }
+
+   const esLintResult = await eslintFilesByExtension(filesByExtension);
+   if (esLintResult.error) {
+      console.log('Please fix these errors before to commit:' + esLintResult.report);
+      console.log('Eslint executed in ' + esLintResult.numberFiles + ' files ');
+      if (codeQualityConfig.eslintStopCommit) {
+         process.exit(1);
+      }
+   } else {
+      console.log('Eslint executed in ' + esLintResult.numberFiles + ' files ');
+   }
+};
+
+/**
+ * Run `prettier` and `eslint` on changed files before commit;
+ * You can also call this function using: npm run code-quality
+ */
+const preCommit = async () => {
+   const files = await getChangedFiles();
+   if (!files) {
+      console.log("Couldn't find any file that hand been changed");
+      return;
+   }
+   console.log('File to be analyzed before commit:\n' + files.join('\n'));
+   const filesByExtension = groupFilesByExtension(files);
+   await preCommitPrettier(filesByExtension);
+   await preCommitEslint(filesByExtension);
+};
+
+/**
+ * Execute ESlint analysis and try to fix problems
+ * @param filesByExtension: we pass a Map json object where the key is the extension, value is an Array with absolute file paths
+ * @returns {Promise<{report: string, numberFiles: number, error: boolean}>}
+ */
+const eslintFilesByExtension = async filesByExtension => {
+   let error = false;
+   let report = '';
+   let numberFiles = 0;
+   const extensions = Object.keys(filesByExtension);
+   for (const extension of extensions.filter(extension => extension === 'ts' || extension === 'js')) {
+      const esLintFiles = filesByExtension[extension];
+      numberFiles += esLintFiles.length;
+      const eslintResult = await eslintFiles(esLintFiles);
+      error = error || eslintResult.error;
+      if (!!eslintResult.report && eslintResult.report.trim().length > 0) {
+         report += eslintResult.report;
+      }
+   }
+
+   return { error, report, numberFiles };
+};
+
+/**
+ * Run `prettier` and `eslint` on all project files;
+ * You can also call this function using: npm run full-code-quality
+ */
+const full = async () => {
+   const files = getAllFiles(__dirname);
+   const filesByExtension = groupFilesByExtension(files);
+   await fullPrettier(filesByExtension);
+   console.log("Running EsLint on all files. Please wait...")
+   await fullEslint(filesByExtension);
+};
+
+/**
+ * Run `prettier` on all project files;
+ * You can also call this function using: npm run full-code-quality
+ */
+const fullPrettier = async filesByExtension => {
+   if (!filesByExtension) {
+      const files = getAllFiles(__dirname);
+      filesByExtension = groupFilesByExtension(files);
+   }
+   prettifyFiles(filesByExtension);
+};
+
+/**
+ * Run `eslint` on all project files;
```

Comment:
Following the comments from Slack, I think it would be best to create an initial critical list as a starting point here. The ESLinter generates a lot of errors but most of those are not criticial.
So I would start by creating a list of the most critical folders, mainly:

client source files' folder
core source files' folders
probably also spec.js test files folder on the angular e2e-app.

Reply:
So the fullEslint would only run for all files in these folders only and the whole repo. Definitely we can extend this to include more in a future PR as needed.

Reply:
I changed a bit the script and I created 3 more target scripts:

-"full-code-quality-eslint-core": "node code_quality.js -- mode=full_eslint sourcePaths=core/src,core/test report=core_full_eslint_report.html",
-"full-code-quality-eslint-client": "node code_quality.js -- mode=full_eslint sourcePaths=client/src report=client_full_eslint_report.html",
-"full-code-quality-eslint-e2e-tests": "node code_quality.js -- mode=full_eslint sourcePaths=test/e2e-test-application/e2e/tests report=e2e-test_full_eslint_report.html"

Via command line, you can run eslint in any specific path(s) and save the report with a specific name....
For example this is the command for core eslint

node code_quality.js -- mode=full_eslint sourcePaths=core/src,core/test report=core_full_eslint_report.html